packages/gatsby/src/services/calculate-dirty-queries.ts

```diff
@@ -5,11 +5,48 @@ import { assertStore } from "../utils/assert-store"
 
 export async function calculateDirtyQueries({
   store,
+  websocketManager,
+  currentlyHandledPendingQueryRuns,
 }: Partial<IQueryRunningContext>): Promise<{
   queryIds: IGroupedQueryIds
 }> {
   assertStore(store)
   const state = store.getState()
   const queryIds = calcDirtyQueryIds(state)
-  return { queryIds: groupQueryIds(queryIds) }
+
+  let queriesToRun: Array<string> = queryIds
+
+  if (
+    process.env.gatsby_executing_command === `develop` &&
+    process.env.GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND
+  ) {
+    // 404 are special cases in our runtime that ideally use
+    // generic things to work, but for now they have special handling
+    const pagePathFilter = new Set([`/404.html`, `/dev-404-page/`])
+
+    // we want to make sure we run queries for pages that user currently
+    // view in the browser
+    if (websocketManager?.activePaths) {
+      websocketManager.activePaths.forEach(
+        pagePathFilter.add.bind(pagePathFilter)
+      )
```

Comment:
Nit: a little bit hard to read

```suggestion
      for (const activePath of websocketManager.activePaths) {
        pagePathFilter.add(activePath)
      }
```